### PR TITLE
Uplift menu to consume popup updates

### LINKED ISSUE #897: Uplift Menu to consume popup with hookshot
# Issue Description
Uplift Menu to consume popup with hookshot.

## Issue Type

- [ ] New Feature
- [x] Enhancement
- [ ] Bug
- [ ] Other

## Expected Behavior
This should allow a menu to be dynamically sized by it's content. Also, the menu should place first focus on the first menu item.

## Current Behavior
Menu is sized with a best guess of height based on number of items and given one of the limited popup height options. The menu also currently will focus on the popup container first on tab, before focusing on the first menu item.

### PR DESCRIPTION
**Pull request by jmsv6d**
### Summary
Resolves #897. Uplifts menu to consume popup updates. 
* Fixes issue where menu container would get focus before content could be focused
* Updated to use popup auto sizing for height
* Added option for auto width sizing in menu and collapsible menu view.

Notes: In order for new layout needs to work slide group could no longer be used (slide group slides have a height of 100% which would collapse the content). I also added logic such that the menu would retain the height and width of the main menu when the menu would navigate to a submenu. 

Deployment Link: https://terra-core-deployed-pr-942.herokuapp.com/static/#/site/menu

### Additional Details
If you have anything else that you think may be relevant to this issue, list it here. Additional information can help us better understand your changes and speed up the review process.

Please add your name to the [CONTRIBUTORS.md] file. Adding your name to the [CONTRIBUTORS.md] file signifies agreement to all rights and reservations provided by the [License].

Thanks for contributing to Terra. 
@cerner/terra

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

### CODE REVIEW

File: packages/terra-collapsible-menu-view/src/CollapsibleMenuView.jsx [hunk 1]
```diff
@@ -21,9 +21,9 @@ const propTypes = {
 
   /**
    * A string representation of the width in px, limited to:
-   * 160, 240, 320, 640, 960, 1280, 1760
+   * 160, 240, 320, 640, 960, 1280, 1760, or auto
    */
-  menuWidth: PropTypes.oneOf(['160', '240', '320', '640', '960', '1280', '1760']),
+  menuWidth: PropTypes.oneOf(['160', '240', '320', '640', '960', '1280', '1760', 'auto']),
```
Comment (dkasper-was-taken) [inline packages/terra-collapsible-menu-view/src/CollapsibleMenuView.jsx]: Should be able to use:
```
menuWidth: PropTypes.oneOf(Object.keys(Popup.Opts.widths));
```

File: packages/terra-collapsible-menu-view/src/CollapsibleMenuViewItem.jsx [hunk 2]
```diff
@@ -62,9 +62,9 @@ const propTypes = {
 
   /**
    * A string representation of the width in px, limited to:
-   * 160, 240, 320, 640, 960, 1280, 1760
+   * 160, 240, 320, 640, 960, 1280, 1760, or auto
    */
-  menuWidth: PropTypes.oneOf(['160', '240', '320', '640', '960', '1280', '1760']),
+  menuWidth: PropTypes.oneOf(['160', '240', '320', '640', '960', '1280', '1760', 'auto']),
```
Comment (dkasper-was-taken) [inline packages/terra-collapsible-menu-view/src/CollapsibleMenuViewItem.jsx]: Should be able to use:
```
menuWidth: PropTypes.oneOf(Object.keys(Popup.Opts.widths));
```

File: packages/terra-menu/src/_MenuContent.jsx [hunk 3]
```diff
@@ -66,13 +106,40 @@ class MenuContent extends React.Component {
...
     if (event.nativeEvent.keyCode === MenuUtils.KEYCODES.ENTER || event.nativeEvent.keyCode === MenuUtils.KEYCODES.SPACE) {
       event.preventDefault();
       this.props.onRequestBack();
     }
   }
 
+  validateFocus(node) {
+    if (this.needsFocus && node) {
+      node.focus();
+      this.needsFocus = document.activeElement !== node;
```
Comment (kschuste) [inline packages/terra-menu/src/_MenuContent.jsx]: when would this ever not be false?
the previous line you set focus on the node making it the active element correct?

File: packages/terra-menu/src/Menu.jsx [hunk 4]
```diff
@@ -66,42 +63,31 @@ const defaultProps = {
...
-      if (child.props.children && child.props.children.length > 0) {
-        itemCount += child.props.children.length;
-      } else if (child.type === <MenuDivider />.type) {
-        dividerCount += 1;
-      } else {
-        itemCount += 1;
+  setPageDimensions(node) {
+    if (node) {
+      this.pageHeight = node.clientHeight;
+      if (this.props.contentWidth === 'dynamic') {
```
Comment (emilyrohrbough) [inline packages/terra-menu/src/Menu.jsx]: This should be auto now 😄

File: packages/terra-menu/src/_MenuContent.jsx (hunk 3 above)
Comment (dkasper-was-taken) [reply packages/terra-menu/src/_MenuContent.jsx]: If the item cannot take focus as a result of being hidden or overlapped, potentially as the result of animation or the timing of the callback. Largely its to guard against the behavior of focus trap, when it sets a list of tabbable items, if for any reason the item is removed from that list, it won't receive focus.

File: packages/terra-site/src/examples/menu/MenuWidths.jsx [hunk 5]
```diff
@@ -26,6 +26,7 @@ class MenuWidths extends React.Component {
             <option key="960" value="960">960</option>
             <option key="1280" value="1280">1280</option>
             <option key="1760" value="1760">1760</option>
+            <option key="auto" value="auto">Dynamic</option>
```
Comment (yuderekyu) [inline packages/terra-site/src/examples/menu/MenuWidths.jsx]: Nit: *Auto

File: packages/terra-menu/src/_MenuContent.jsx [hunk 6]
```diff
@@ -37,10 +37,47 @@ const propTypes = {
...
+   * Indicates if the content should be hidden.
+   */
+  isHidden: PropTypes.bool,
+  /**
+   * Ref callback function to be applied to content container.
+   */
+  refCallback: PropTypes.func,
 };
 
 const defaultProps = {
```
Comment (emilyrohrbough) [inline packages/terra-menu/src/_MenuContent.jsx]: Do we need to add default values of false for isHidden, isHeightBounded, and isWidthBounded?

File: packages/terra-menu/src/_MenuContent.jsx [hunk 7]
```diff
@@ -37,10 +37,47 @@ const propTypes = {
...
+   */
+  boundingRef: PropTypes.func,
+  /**
+   * Indicates if the menu content should set default focus on itself.
+   */
+  isFocused: PropTypes.bool,
+  /**
+   * Indicates if menu's height has been constrained by bounding container.
+   */
+  isHeightBounded: PropTypes.bool,
```
Comment (emilyrohrbough) [inline packages/terra-menu/src/_MenuContent.jsx]: Maybe I'm missing where this is determined, but I don't think Menu ever uses the isHeightBounded and isWidthBounded to the MenuContent component?

File: packages/terra-menu/src/_MenuContent.jsx [hunk 8]
```diff
@@ -202,15 +271,42 @@ class MenuContent extends React.Component {
...
+    const contentClass = cx([
+      'content',
+      { submenu: isSubMenu },
+      { 'hidden-page': this.props.isHidden },
+      { fullscreen: isFullScreen },
+    ]);
 
-    const header = this.buildHeader();
+    let header;
+    if (isFullScreen || isSubMenu) {
```
Comment (emilyrohrbough) [inline packages/terra-menu/src/_MenuContent.jsx]: Which test example should we view to see a fullscreen menu?

File: packages/terra-menu/src/_MenuContent.jsx (hunk 7 above)
Comment (jmsv6d) [reply packages/terra-menu/src/_MenuContent.jsx]: These props are added by the Popup.

File: packages/terra-menu/src/_MenuContent.jsx (hunk 8 above)
Comment (jmsv6d) [reply packages/terra-menu/src/_MenuContent.jsx]: The bounded menu is the "fullscreen" menu.

File: packages/terra-menu/src/_MenuContent.jsx (hunk 6 above)
Comment (jmsv6d) [reply packages/terra-menu/src/_MenuContent.jsx]: Added here: https://github.com/cerner/terra-core/pull/942/commits/1c7fac345b5bd12453513ffe3a54b4fd568dc5f5

**Comment by saedar** [comment]
In the site examples, it looks like some disabled menu items are still selectable.

screen shot 2017-10-24 at 9 47 57 am

Is that desired?

**Comment by jmsv6d** [comment]
@saedar That issue looks to be unrelated to the changes in this pr (its occurring on the engineering site as well). But I think it would be a pretty simple fix so I think I could include it with these changes.

File: packages/terra-collapsible-menu-view/CHANGELOG.md [hunk 9]
```diff
@@ -3,6 +3,8 @@ ChangeLog
 
 Unreleased
 -----------------
+### Added
+* Added option for auto width sizing.
```
Comment (emilyrohrbough) [inline packages/terra-collapsible-menu-view/CHANGELOG.md]: Update to include your disable fix

File: packages/terra-menu/src/Menu.jsx [hunk 10]
```diff
@@ -66,42 +63,31 @@ const defaultProps = {
...
-    let dividerCount = 0;
-
-    React.Children.forEach(this.props.children, (child) => {
-      if (child.props.children && child.props.children.length > 0) {
-        itemCount += child.props.children.length;
-      } else if (child.type === <MenuDivider />.type) {
-        dividerCount += 1;
-      } else {
-        itemCount += 1;
+  setPageDimensions(node) {
```
Comment (emilyrohrbough) [inline packages/terra-menu/src/Menu.jsx]: Is `setPageDimensions` setting the menu container's height & width?

File: packages/terra-menu/src/Menu.jsx [hunk 11]
```diff
@@ -168,18 +149,20 @@ class Menu extends React.Component {
...
+        classNameContent={classNameContent}
         classNameOverlay={classNameOverlay}
         isOpen={isOpen}
         onRequestClose={onRequestClose}
         targetRef={targetRef}
         isHeaderDisabled
+        isContentFocusDisabled
       >
-        <SlideGroup items={slides} />
+        {slides}
```
Comment (emilyrohrbough) [inline packages/terra-menu/src/Menu.jsx]: So in conjunction with my comment above about `setPageDimensions`, maybe rename slides to pages to match context

**Comment by emilyrohrbough** [comment]
non-selectable items are still applying selectable styles:

screen shot 2017-10-26 at 3 04 43 pm

File: packages/terra-menu/src/_MenuContent.jsx [hunk 12]
```diff
@@ -66,13 +109,40 @@ class MenuContent extends React.Component {
...
+    if (this.props.refCallback) {
+      this.props.refCallback(node);
+    }
+    this.contentNode = node;
+    this.validateFocus(node);
+  }
+
   isSelectable() {
     let isSelectable = false;
     React.Children.forEach(this.props.children, (child) => {
```
Comment (emilyrohrbough) [inline packages/terra-menu/src/_MenuContent.jsx]: Do we need to check for isDisabled here as well?

**Comment by emilyrohrbough** [comment]
How does disabled sub-menu items present?

**Comment by jmsv6d** [comment]
@emilyrohrbough That is intentional behavior. The only menu items that don't get hover styling are the disabled ones. I'm really not sure what the disabled sub-menu items look like. The css style is opacity: 0.3. I can send you a screenshot, but that's really unrelated to the changes in this pull request.

**Comment by emilyrohrbough** [comment]
Add tests for auto sizing & for disabled items.

**Comment by mmalaker** [comment]
Functional verification completed. The menu examples display and function correctly after consuming popup updates.

**Pull request merged**